caput/profile.py

```diff
@@ -163,3 +165,308 @@ def __exit__(self, *args, **kwargs):
+        use_profiler: bool = True,
+        label: str = "",
+        logger: Optional[logging.Logger] = None,
+        comm: Optional["mpiutil.MPI.IntraComm"] = None,
+        path: Optional[os.PathLike] = None,
+    ):
+        self._use_profiler = use_profiler
+        self._label = label
+        self._usage = {}
+        self._logger = logger
+        self.comm = comm
+
+        if self.comm is None:
+            rank = mpiutil.rank
+        else:
+            rank = self.comm.rank
+
+        if path is None:
+            self.path = Path.cwd()
+        else:
+            if not self.path.is_dir() or not self.path.exists():
+                raise ValueError(
+                    f"Make sure {self.path} passed to PSUtillProfiler is a directory that exists."
+                )
+            self.path = Path(path)
+
+        self.path = self.path / f"perf_{rank}.csv"
+
+        self._start_cpu_times = {}
+        self._start_memory = {}
+        self._start_io = {}
+        self._start_disk_io = {}
+        self._start_time = {}
+
+        super().__init__()
+
+        if self._use_profiler and not self.path.exists():
+
+            import csv
+
+            with open(self.path, mode="w") as fp:
+
+                colnames = [
+                    "task_name",
+                    "time_s",
+                    "cpu_times_user",
+                    "cpu_times_system",
+                    "cpu_times_children_user",
+                    "cpu_times_children_system",
+                    "cpu_times_iowait",
+                    "average_cpu_load_percent",
+                    "disk_io_read_count",
+                    "disk_io_write_count",
+                    "disk_io_read_bytes",
+                    "disk_io_write_bytes",
+                    "disk_io_read_chars",
+                    "disk_io_write_chars",
+                    "memory_change_uss",
+                    "current_available_memory_bytes",
+                    "current_total_used_memory_bytes",
+                ]
+                cw = csv.writer(fp)
+                cw.writerow(colnames)
+
+        if self._use_profiler and self._logger:
+            self._logger.info(f"Profiling pipeline: {self.cpu_count} cores available.")
+
+    def __eq__(self, other):
+        if not isinstance(other, psutil.Process):
+            return False
+        return (
+            psutil.Process.__eq__(self, other)
+            and self._start_cpu_times == other._start_cpu_times
+            and self._start_memory == other._start_memory
+            and self._start_io == other._start_io
+            and self._start_disk_io == other._start_disk_io
+            and self._start_time == other._start_time
+        )
+
+    def __enter__(self):
+        if not self._use_profiler:
+            return
+        self.start(self._label)
+
+    def __exit__(self, *args, **kwargs):
+        if not self._use_profiler:
+            return
+        self.stop(self._label)
+
+    def start(self, label):
```

Comment:
This looks good, but I don't understand the point of `label` being used here and in `.stop(...)`. It doesn't change at all, and these PSUtil objects are ephemeral, so why does in need to get passed in, and why is `_start_time` indexed by it?

Reply:
Yeah...it was left-over from when the design was that you would create a single instance of `PSUtilProfiler` and use it for everything. We since moved away from that design, and it would clean things up to modify it. I will clean that up.

Reply:
If we do alter this design, then `PSUtilProfiler` has to be initialised each time someone wants to profile something new!